In [514]:
%pip install anthropic

You should consider upgrading via the '/usr/local/bin/python3.9 -m pip install --upgrade pip' command.
Note: you may need to restart the kernel to use updated packages.


In [515]:
%pip install pdfplumber

You should consider upgrading via the '/usr/local/bin/python3.9 -m pip install --upgrade pip' command.
Note: you may need to restart the kernel to use updated packages.


In [ ]:
import anthropic

client = anthropic.Anthropic(
    api_key=""
)

Chunk the original pdf

In [539]:
import os
import pdfplumber

# === CONFIGURATION ===
input_pdf_folder = "Original pdf/layer3_v11"
output_base_folder = "acts/layer3_v11"
max_tokens = 3000  # Approximate word limit per chunk

# === FUNCTION: Extract text from PDF ===
def extract_text_from_pdf(pdf_path):
    with pdfplumber.open(pdf_path) as pdf:
        text = ""
        for page in pdf.pages:
            page_text = page.extract_text()
            if page_text:
                text += page_text + "\n"
    return text

# === FUNCTION: Chunk text ===
def chunk_text(text, max_tokens=3000):
    chunks = []
    current_chunk = []
    current_length = 0

    paragraphs = [p.strip() for p in text.split("\n") if p.strip()]

    for paragraph in paragraphs:
        paragraph_length = len(paragraph.split())
        if paragraph_length > max_tokens:
            words = paragraph.split()
            for i in range(0, len(words), max_tokens):
                sub_paragraph = " ".join(words[i:i + max_tokens])
                chunks.append(sub_paragraph)
        elif current_length + paragraph_length > max_tokens:
            chunks.append("\n\n".join(current_chunk))
            current_chunk = []
            current_length = 0
        current_chunk.append(paragraph)
        current_length += paragraph_length

    if current_chunk:
        chunks.append("\n\n".join(current_chunk))

    return chunks

# === FUNCTION: Save chunks ===
def save_chunks_to_txt(chunks, output_dir):
    os.makedirs(output_dir, exist_ok=True)
    for i, chunk in enumerate(chunks):
        chunk_file_path = os.path.join(output_dir, f"chunk_{i + 1}.txt")
        with open(chunk_file_path, "w", encoding="utf-8") as f:
            f.write(chunk)
    print(f"✅ Saved {len(chunks)} chunks to '{output_dir}'")

# === MAIN PROCESS ===
for filename in os.listdir(input_pdf_folder):
    if filename.lower().endswith(".pdf"):
        act_name = filename.replace(".pdf", "").strip()
        input_path = os.path.join(input_pdf_folder, filename)
        output_dir = os.path.join(output_base_folder, act_name)

        print(f"\n📄 Processing: {act_name}")
        raw_text = extract_text_from_pdf(input_path)
        text_chunks = chunk_text(raw_text, max_tokens=max_tokens)
        save_chunks_to_txt(text_chunks, output_dir)

print("\n🎉 All PDFs have been processed and chunked.")

CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox



📄 Processing: Electoral Integrity Amendment Bill


CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, def

✅ Saved 1 chunks to 'acts/layer3_v11/Electoral Integrity Amendment Bill'

📄 Processing: Accident Rehabilitation and Compensation Insurance act 1992


CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, def

✅ Saved 16 chunks to 'acts/layer3_v11/Accident Rehabilitation and Compensation Insurance act 1992'

📄 Processing: Child Support Amendment Act 1993
✅ Saved 1 chunks to 'acts/layer3_v11/Child Support Amendment Act 1993'

📄 Processing: Evidence Act 1908


CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, def

✅ Saved 5 chunks to 'acts/layer3_v11/Evidence Act 1908'

📄 Processing: Education Amendment Act 1989


CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, def

✅ Saved 10 chunks to 'acts/layer3_v11/Education Amendment Act 1989'

📄 Processing: Care of Children Amendment Act 2005


CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox


✅ Saved 1 chunks to 'acts/layer3_v11/Care of Children Amendment Act 2005'

📄 Processing: Child Support Amendment Act 1994


CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox


✅ Saved 1 chunks to 'acts/layer3_v11/Child Support Amendment Act 1994'

📄 Processing: COVID19 Response Taxation and Other Regulatory Urgent Measures Act 2020


CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, def

✅ Saved 3 chunks to 'acts/layer3_v11/COVID19 Response Taxation and Other Regulatory Urgent Measures Act 2020'

📄 Processing: Arms Amendment Act 1992


CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, def

✅ Saved 2 chunks to 'acts/layer3_v11/Arms Amendment Act 1992'

📄 Processing: Employment Contracts act 1991


CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, def

✅ Saved 11 chunks to 'acts/layer3_v11/Employment Contracts act 1991'

📄 Processing: Accident Insurance Act 1998


CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, def

✅ Saved 38 chunks to 'acts/layer3_v11/Accident Insurance Act 1998'

📄 Processing: Aged and Infirm Persons Protection Act 1912


CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox


✅ Saved 2 chunks to 'acts/layer3_v11/Aged and Infirm Persons Protection Act 1912'

📄 Processing: Births Deaths Marriages and Relationships Registration Act 1995
✅ Saved 13 chunks to 'acts/layer3_v11/Births Deaths Marriages and Relationships Registration Act 1995'

📄 Processing: COVID19 Response Vaccinations Legislation Act 2021


CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, def

✅ Saved 2 chunks to 'acts/layer3_v11/COVID19 Response Vaccinations Legislation Act 2021'

📄 Processing: Arms Amendment Act 1999
✅ Saved 1 chunks to 'acts/layer3_v11/Arms Amendment Act 1999'

📄 Processing: Acts and Regulations Publication Act 1989


CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox


✅ Saved 1 chunks to 'acts/layer3_v11/Acts and Regulations Publication Act 1989'

📄 Processing: Arms Legislation Act 2020
✅ Saved 13 chunks to 'acts/layer3_v11/Arms Legislation Act 2020'

📄 Processing: Arms Act 1983


CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, def

✅ Saved 23 chunks to 'acts/layer3_v11/Arms Act 1983'

📄 Processing: Child Support Amendment Act 2006


CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, def

✅ Saved 5 chunks to 'acts/layer3_v11/Child Support Amendment Act 2006'

📄 Processing: Animals Protection Act 1960


CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox


✅ Saved 2 chunks to 'acts/layer3_v11/Animals Protection Act 1960'

📄 Processing: Evidence Act 2006


CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, def

✅ Saved 21 chunks to 'acts/layer3_v11/Evidence Act 2006'

📄 Processing: Arms Amendment Act 2000
✅ Saved 1 chunks to 'acts/layer3_v11/Arms Amendment Act 2000'

📄 Processing: Child Support Amendment Act 1998
✅ Saved 1 chunks to 'acts/layer3_v11/Child Support Amendment Act 1998'

📄 Processing: Evidence Amendment Act No 2 1980


CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, defaulting to MediaBox
CropBox missing from /Page, def

✅ Saved 3 chunks to 'acts/layer3_v11/Evidence Amendment Act No 2 1980'

🎉 All PDFs have been processed and chunked.


In [464]:
# import os
# import pdfplumber

# # Function to extract text from a PDF (all pages)
# def extract_text_from_pdf(pdf_path):
#     with pdfplumber.open(pdf_path) as pdf:
#         text = ""
#         for page in pdf.pages:
#             page_text = page.extract_text()
#             if page_text:
#                 text += page_text + "\n"
#     return text

# # Function to chunk text into smaller pieces
# def chunk_text(text, max_tokens=3000):
#     chunks = []
#     current_chunk = []
#     current_length = 0

#     paragraphs = [p.strip() for p in text.split("\n") if p.strip()]  # Clean up blank lines

#     for paragraph in paragraphs:
#         paragraph_length = len(paragraph.split())
#         if paragraph_length > max_tokens:
#             words = paragraph.split()
#             for i in range(0, len(words), max_tokens):
#                 sub_paragraph = " ".join(words[i:i + max_tokens])
#                 chunks.append(sub_paragraph)
#         elif current_length + paragraph_length > max_tokens:
#             chunks.append("\n\n".join(current_chunk))
#             current_chunk = []
#             current_length = 0
#         current_chunk.append(paragraph)
#         current_length += paragraph_length

#     if current_chunk:
#         chunks.append("\n\n".join(current_chunk))

#     return chunks

# # Function to save chunks to individual .txt files
# def save_chunks_to_txt(chunks, output_dir):
#     os.makedirs(output_dir, exist_ok=True)
#     for i, chunk in enumerate(chunks):
#         chunk_file_path = os.path.join(output_dir, f"chunk_{i + 1}.txt")
#         with open(chunk_file_path, "w", encoding="utf-8") as chunk_file:
#             chunk_file.write(chunk)
#         print(f"Saved chunk {i + 1} to {chunk_file_path}")

# # Example usage
# pdf_path = "Original pdf/Animal Welfare Act 1999.pdf"  # Replace with your PDF file
# output_directory = "pdf_chunks"

# raw_text = extract_text_from_pdf(pdf_path)
# text_chunks = chunk_text(raw_text, max_tokens=3000)
# save_chunks_to_txt(text_chunks, output_directory)

# print(f"PDF text split into {len(text_chunks)} chunks and saved to '{output_directory}'.")

Function to call Claud 3 Haiku Pipeline

In [ ]:
import re
import json
import anthropic
import os

if not client.api_key:
    raise ValueError("🚨 ANTHROPIC_API_KEY is not set! Configure it before running the script.")

def extract_acts_relationships_claude(chunk_text, source_act):
    
    prompt = f"""
    Analyze the following legal text and extract all **external** acts and their relationships.

    **Rules:**
    - Only extract **external references** (ignore internal references to the same act).
    - Standardize act names (e.g., "1995 Act" → "Domestic Violence Act 1995").
    - Remove duplicate entries and ensure consistency.

    **Valid Relationship Types:**
    - "Amended" → This act modifies another act.
    - "Repealed" → This act removes another act.
    - "Referenced" → This act only mentions another act but does not modify it.

    **Return the output in strict JSON format:**
    ```json
    {{
      "acts": [
        {{"source": "{source_act}", "target": "Sentencing Act 2002", "relationship": "Referenced"}},
        {{"source": "{source_act}", "target": "Privacy Act 2020", "relationship": "Amended"}}
      ]
    }}
    ```

    Legal Text:
    {chunk_text}
    """

    try:
        response = client.messages.create(
            model="claude-3-haiku-20240307",
            max_tokens=3000,
            temperature=0.2,
            messages=[
                {"role": "user", "content": prompt}
            ]
        )

        raw_response = response.content[0].text.strip()
        print(f"📝 Raw API Response:\n{raw_response[:400]}...")

        # ✅ Extract JSON part using regex
        match = re.search(r"\{[\s\S]*\}", raw_response)
        if match:
            json_text = match.group(0)
            return json.loads(json_text)
        else:
            print("⚠️ No JSON block found in response.")
            return {"acts": []}

    except Exception as e:
        print(f"🚨 Error calling Claude or parsing response: {e}")
        return {"acts": []}

In [541]:
# === MAIN AUTOMATION ===

base_folder = "acts/layer3_v11"
output_file = "layer_2_relationships_appended.json"
all_results = []

# Loop through each act folder
for act_name in sorted(os.listdir(base_folder)):
    act_folder_path = os.path.join(base_folder, act_name)
    if not os.path.isdir(act_folder_path):
        continue

    print(f"\nProcessing act: {act_name}")

    # Loop through each chunk
    for file in sorted(os.listdir(act_folder_path)):
        if file.endswith(".txt"):
            with open(os.path.join(act_folder_path, file), "r", encoding="utf-8") as f:
                text = f.read().strip()
                if text:
                    result = extract_acts_relationships_claude(text, source_act=act_name)
                    if result:
                        all_results.append(result)

# Load existing results
if os.path.exists(output_file):
    with open(output_file, "r", encoding="utf-8") as f:
        existing_data = json.load(f)
else:
    existing_data = []

# Combine and save
combined_data = existing_data + all_results

with open(output_file, "w", encoding="utf-8") as f:
    json.dump(combined_data, f, indent=2)

print(f"\n✅ All relationships saved to: {output_file}")


Processing act: Accident Insurance Act 1998
📝 Raw API Response:
Here is the extracted information in the requested JSON format:

```json
{
  "acts": [
    {
      "source": "Accident Insurance Act 1998",
      "target": "Sentencing Act 2002",
      "relationship": "Referenced"
    },
    {
      "source": "Accident Insurance Act 1998",
      "target": "Privacy Act 2020",
      "relationship": "Amended"
    }
  ]
}
```...
📝 Raw API Response:
Based on the provided legal text, here is the extracted information about external acts and their relationships:

```json
{
  "acts": [
    {
      "source": "Accident Insurance Act 1998",
      "target": "Sentencing Act 2002",
      "relationship": "Referenced"
    },
    {
      "source": "Accident Insurance Act 1998",
      "target": "Privacy Act 2020",
      "relationship": "Amended"
    },
  ...
🚨 Error calling Claude or parsing response: Expecting ',' delimiter: line 32 column 6 (char 831)
📝 Raw API Response:
Here is the JSON output with the 

In [ ]:
# import os
# import json
# #from your_module import extract_acts_relationships_claude  # Update if needed

# folder_path = "acts/Income Tax Act 2007"
# output_file = "layer_2_relationships_appended.json"
# source_act = "Income Tax Act 2007"
# all_results = []

# for file in sorted(os.listdir(folder_path)):
#     if file.endswith(".txt"):
#         with open(os.path.join(folder_path, file), "r", encoding="utf-8") as f:
#             text = f.read().strip()
#             if text:
#                 result = extract_acts_relationships_claude(text, source_act=source_act)
#                 if result:
#                     all_results.append(result)

# # Load existing data if present
# if os.path.exists(output_file):
#     with open(output_file, "r", encoding="utf-8") as f:
#         existing_data = json.load(f)
# else:
#     existing_data = []

# # Combine and save
# combined_data = existing_data + all_results

# with open(output_file, "w", encoding="utf-8") as f:
#     json.dump(combined_data, f, indent=2)

# print("✅ All relationships saved to:", output_file)

📝 Raw API Response:
Here is the JSON output with the extracted external acts and their relationships:

```json
{
  "acts": [
    {
      "source": "Income Tax Act 2007",
      "target": "Sentencing Act 2002",
      "relationship": "Referenced"
    },
    {
      "source": "Income Tax Act 2007",
      "target": "Privacy Act 2020",
      "relationship": "Amended"
    },
    {
      "source": "Income Tax Act 2007",
    ...
📝 Raw API Response:
Based on the provided legal text, there are no external acts referenced. The text appears to be focused on internal provisions and calculations within the Income Tax Act 2007, without any references to other external acts.

Therefore, the output in the requested JSON format would be:

```json
{
  "acts": []
}
```...
📝 Raw API Response:
Based on the provided legal text, the following external acts and their relationships can be extracted:

```json
{
  "acts": [
    {
      "source": "Customs and Excise Act 2018",
      "target": "Income Tax Act 2007",

Clean the json output

In [542]:
import json

# Load the original file
with open("layer_2_relationships_appended.json", "r", encoding="utf-8") as f:
    raw_data = json.load(f)

flat_relationships = []

# Flatten all "acts" entries into one list
for entry in raw_data:
    acts = entry.get("acts", [])
    flat_relationships.extend(acts)

# Remove duplicates based on (source, target, relationship)
unique_relationships = []
seen = set()

for item in flat_relationships:
    # Skip incomplete entries
    if not all(k in item for k in ("source", "target", "relationship")):
        print(f"⚠️ Skipped incomplete entry: {item}")
        continue

    key = (item["source"], item["target"], item["relationship"])
    if key not in seen:
        seen.add(key)
        unique_relationships.append(item)

# Save cleaned list to a new file
with open("layer_2_relationships_cleaned.json", "w", encoding="utf-8") as f:
    json.dump(unique_relationships, f, indent=2)

print(f"Cleaned and deduplicated file saved to 'layer_2_relationships_cleaned.json'")
print(f"Total unique relationships: {len(unique_relationships)}")

⚠️ Skipped incomplete entry: {'source': 'Sentencing Act 2002', 'relationship': 'Referenced'}
⚠️ Skipped incomplete entry: {'source': 'Privacy Act 2020', 'relationship': 'Referenced'}
⚠️ Skipped incomplete entry: {'source': 'Companies Act 1993', 'relationship': 'Referenced'}
⚠️ Skipped incomplete entry: {'source': 'Land Transfer Act 1952', 'relationship': 'Referenced'}
⚠️ Skipped incomplete entry: {'source': 'Deeds Registration Act 1908', 'relationship': 'Referenced'}
⚠️ Skipped incomplete entry: {'source': 'Crown Minerals Act 1991', 'relationship': 'Referenced'}
⚠️ Skipped incomplete entry: {'source': 'Education Act 1964', 'relationship': 'Repealed'}
⚠️ Skipped incomplete entry: {'source': 'Education Act 1989', 'relationship': 'Repealed'}
⚠️ Skipped incomplete entry: {'source': 'State Sector Act 1988', 'relationship': 'Repealed'}
⚠️ Skipped incomplete entry: {'source': 'Education Standards Act 2001', 'relationship': 'Repealed'}
⚠️ Skipped incomplete entry: {'source': 'Domestic Violence

Get a count of acts

In [543]:
from collections import defaultdict

# Load the cleaned file
with open("layer_2_relationships_cleaned.json", "r", encoding="utf-8") as f:
    data = json.load(f)

# Count total relationships
total_relationships = len(data)

# Track all unique act names (source or target)
unique_acts = set()

# Track number of relationships per act (as source OR target)
act_relationship_count = defaultdict(int)

for entry in data:
    source = entry["source"]
    target = entry["target"]

    unique_acts.add(source)
    unique_acts.add(target)

    act_relationship_count[source] += 1
    act_relationship_count[target] += 1

# Print overall stats
print(f"Total relationships: {total_relationships}")
print(f"Unique acts involved (source or target): {len(unique_acts)}")
print(f"\nNumber of relationships per act (source or target):")

for act, count in sorted(act_relationship_count.items(), key=lambda x: x[1], reverse=True):
    print(f"  - {act}: {count}")

Total relationships: 9090
Unique acts involved (source or target): 2015

Number of relationships per act (source or target):
  - Privacy Act 2020: 984
  - Sentencing Act 2002: 859
  - None: 367
  - Income Tax Act 2007: 320
  - Legislation Act 2019: 252
  - Criminal Procedure Act 2011: 235
  - Companies Act 1993: 168
  - District Court Act 2016: 165
  - Domestic Violence Act 1995: 147
  - Education and Training Act 2020: 136
  - Income Tax Act 2004: 134
  - Land Transport Act 1998: 132
  - Resource Management Act 1991: 124
  - Employment Relations Act 2000: 122
  - Tax Administration Act 1994: 121
  - Public Service Act 2020: 120
  - Crimes Act 1961: 114
  - Local Government Act 2002: 114
  - Electoral Act 1993: 111
  - Education Act 1989: 108
  - Social Security Act 2018: 103
  - Social Security Act 1964: 102
  - Land Transfer Act 2017: 99
  - Search and Surveillance Act 2012: 97
  - Summary Proceedings Act 1957: 96
  - Official Information Act 1982: 91
  - Evidence Act 2006: 91
  - Cr

In [469]:
print("\nList of unique acts:")
for act in sorted([a for a in unique_acts if a is not None]):
    print(f"- {act}")


List of unique acts:
- 1947 No 16
- 1964 No 135
- 1964 No 136
- 1977 No 49
- 1985 No 120
- 1988 Act
- 1989 Act
- 1989 No 80
- 1991 No 22
- 1992 Act
- 1992 No 55
- 1995 Act
- 2004 Transfer Notice
- 2006 Transfer Notice
- 2013 Amendment Act
- 2018 Act
- A New Tax System (Wine Equalisation Tax) Act 1999 (Aust)
- AS 1940:2017
- AS 2030.5—2009
- AS 3780-2008
- AS 3848.2-1999
- AS/NZS 1596—2014
- AS/NZS 1850:2009
- AS/NZS 2299.1:2015
- AS/NZS 3833:2007
- AS/NZS 4114.1:2003
- AS/NZS 4452:1997
- AS/NZS 5026:2012
- Abolition of Provinces Act 1875
- Accident Compensation Act (weekly compensation)
- Accident Compensation Act 1982
- Accident Compensation Act 2001
- Accident Insurance Act 1998
- Accident Rehabilitation and Compensation Insurance Act 1992
- Act
- Acts Interpretation Act 1901
- Acts Interpretation Act 1908
- Acts Interpretation Act 1915
- Acts Interpretation Act 1924
- Acts and Regulations Publication Act 1989
- Administration Act 1969
- Administration of Community Sentences and Ord

Find the newly added acts

In [470]:
# Load the cleaned relationship data
with open("layer_2_relationships_cleaned.json", "r", encoding="utf-8") as f:
    data = json.load(f)

# Step 1: Collect all unique acts in the network (source + target)
unique_acts = set()
for entry in data:
    source = entry.get("source")
    target = entry.get("target")
    if source: unique_acts.add(source)
    if target: unique_acts.add(target)

# Step 2: Get list of chunked act folders under acts/
chunked_act_folders = set(os.listdir("acts"))

# Step 3: Find acts that are in the network but not chunked
not_chunked_acts = sorted([act for act in unique_acts if act not in chunked_act_folders])

# Step 4: Output
print(f"📄 Total acts in the network: {len(unique_acts)}")
print(f"📁 Total acts already chunked: {len(chunked_act_folders)}")
print(f"❌ Acts in network but NOT chunked: {len(not_chunked_acts)}")
print("\nMissing act folders:")
for act in not_chunked_acts:
    print(f"- {act}")

# Optional: save to a file
with open("missing_acts.txt", "w", encoding="utf-8") as f:
    for act in not_chunked_acts:
        f.write(act + "\n")
print("✅ Missing acts list saved to 'missing_acts.txt'")

📄 Total acts in the network: 841
📁 Total acts already chunked: 61
❌ Acts in network but NOT chunked: 791

Missing act folders:
- 1947 No 16
- 1964 No 135
- 1964 No 136
- 1977 No 49
- 1985 No 120
- 1988 Act
- 1989 Act
- 1989 No 80
- 1991 No 22
- 1992 Act
- 1992 No 55
- 1995 Act
- 2004 Transfer Notice
- 2006 Transfer Notice
- 2013 Amendment Act
- 2018 Act
- A New Tax System (Wine Equalisation Tax) Act 1999 (Aust)
- AS 1940:2017
- AS 2030.5—2009
- AS 3780-2008
- AS 3848.2-1999
- AS/NZS 1596—2014
- AS/NZS 1850:2009
- AS/NZS 2299.1:2015
- AS/NZS 3833:2007
- AS/NZS 4114.1:2003
- AS/NZS 4452:1997
- AS/NZS 5026:2012
- Abolition of Provinces Act 1875
- Accident Compensation Act (weekly compensation)
- Accident Compensation Act 1982
- Accident Compensation Act 2001
- Accident Insurance Act 1998
- Accident Rehabilitation and Compensation Insurance Act 1992
- Act
- Acts Interpretation Act 1901
- Acts Interpretation Act 1908
- Acts Interpretation Act 1915
- Acts Interpretation Act 1924
- Acts and R

layer 1 total acts: 62

layer 2 total acts: 842, relationships: 2646

layer 3 total acts: 2015   , relationships: 9090
